Adds Ability to Search Active Config Profile in MRTK Inspector

Railboy opened the pull request:
## Overview (Revised)
Profiles can now be searched for sub-profiles and fields by keyword.

![SearchFieldSimple](https://user-images.githubusercontent.com/9789716/66428131-5168d080-e9ca-11e9-8ed8-5860f26fb97a.PNG)

If keywords are entered, individual fields are displayed. By default a field's name, tooltip and (if an object reference) value name are searched.

![SearchResultsSimple](https://user-images.githubusercontent.com/9789716/66428108-4a41c280-e9ca-11e9-9356-04c82bd1d1a4.PNG)

Search can be configured via the Options foldout.

![OptionsSimple](https://user-images.githubusercontent.com/9789716/66428097-4615a500-e9ca-11e9-96fe-339a681e17ad.PNG)

Searches are performed by the `MixedRealitySearchUtility`. This utility can be used to search individual profiles, not just the active configuration profile.

## Changes
- Fixes: #4243, #6225

## Verification
Try searching for typically oft-needed-but-difficult-to-find fields in an active profile and see if the search returns useful results.

> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

Railboy commented:
>Also I wonder, how is this different from just searching in the assets search box? I suppose the main difference is the new window searches just the active profile.

For one example - I constantly forget where to change the hand input simulation settings.

In the search box, I can type 'simulation' and I get this, which isn't terribly illuminating:

![SearchResults](https://user-images.githubusercontent.com/9789716/66143592-45d96c00-e5bc-11e9-9a6c-0288351d3f44.PNG)

With the search system, typing 'simulation' gives me all the simulation-related fields in my active profile, which I can then immediately modify. (See screencap in original post.)

julenka commented:
Following one of our guidlines of ["Keep pull requests small"](https://user-images.githubusercontent.com/9789716/66083040-576d3600-e520-11e9-8186-ca50848fff05.PNG)

>Smaller PRs are reviewed more quickly and thoroughly, are less likely to introduce bugs, easier to roll back, and easier to merge.
> Pull requests should be small enough that an engineer could review it in under 30 minutes. Try to make a minimal change that addresses just one thing. If you must create a large PR, split it into several PRs that go into either your local branch, or a feature branch of MRTK. Avoid adding new assets (e.g. fbx, obj files) and instead aim to re-use existing assets.

I wonder if this is something we could iterate on in a separate branch, with smaller changes? For example, [in your repository ](https://github.com/Railboy/MixedRealityToolkit-Unity/tree/mrtk_searchable_profiles)?

Perhaps we could iterate on the UI / design in that branch with an MRTK dev and also a designer, say @cre8ivepark?

For example, the list of subjects could be very useful, but on first impression feels a little overwhelming:

![image](https://user-images.githubusercontent.com/168492/66147662-21818d80-e5c4-11e9-899e-99be2e8dbfbb.png)

Perhaps maybe cre8ivepark has some suggestions? @ritijain  is also somebody who has worked with designers regarding profiles and can echo / represent the feedback on profiles. @CDiaz-MS  might also have suggestions from improving the usability of profile searching.

Love the intention and direction of this (make profiles searchable), looking forward to iterating with you on this.

julenka commented:
> > Also I wonder, how is this different from just searching in the assets search box? I suppose the main difference is the new window searches just the active profile.
> 
> For one example - I constantly forget where to change the hand input simulation settings.
> 
> In the asset search box, I can type 'simulation' and I get this, which isn't terribly illuminating:
> 
> ![SearchResults](https://user-images.githubusercontent.com/9789716/66143592-45d96c00-e5bc-11e9-9a6c-0288351d3f44.PNG)
> 
> With the search system, typing 'simulation' gives me all the simulation-related fields in my active profile, which I can then immediately modify. (See screencap in original post.)

Fair enough. It might be good to include this motivation in the PR description. Like a "before / after" type motivation.

Railboy commented:
>This PR is right on the border: 35 files, with one file at 400 lines. But the bigger reason I think the PR is "large" is because it's making a significant UI editor change to profiles.

@julenka Most of the changes are just adding a 'Subject' tag to the profiles. If I remove those changes the number of touched files drops to 4.

Since the initial feedback is that subject tags are overwhelming why don't I remove subject tags altogether and stick with the keyword search?

Troy-Ferrell commented:
Love this!

I agree with removing the subject fields. They are a bit overwhelming and don't scale well. New items need to be aware of the Subject tag and then implement it themselves. Now and eventually as the list grows, one would spend more times finding and selecting items in the word space then just typing in the search bar.

Feedback suggestions on UI:
1. Remove subject tags (obviously)
1. Remove "Enter keyword "Helpbox" to just a bolded title like "Search Keywords" or "Search profiles for" etc.
1. Make "options" to "Search Preferences". Also make show true default. so people are aware, it's a little hidden atm
1. Make Results bigger title like in Interactable or at least bolded
1. I haven't dug deep into the code yet, but searching "Hand tracking" brings up results for the camera simulation like "Move vertical"? How is that being matched?
1. The field rendering is a bit hard to digest. It's not always clear what part or what profile a field is part of since you are just getting the one field title. For example, if I was to show the enabled property in a profile...enabled of what? Furthermore, in my screenshots example, the Controller Mapping options shows a lot of elements and its very hard to open, inspector, understand, and find the different fields.

Proposal: maybe a simpler first step approach might just be to list the profiles with relevant results? And then render that entire profile using the Config.RenderProfile() api. It would be awesome if we could render the sub-profile and then bold or mark in red highlight the fields that match, but not sure how we could do that easily. At least I am filtering my profile/results to the areas of interest and then I can see the entire profile with context around that profile. (also, each helpbox profile item section would need a toggle group to show/hide)

Also when I tried to modify one of hte properties found, then I drilled into the actual profile object, it didn't update the corresponding field? (like setting handedness in the pointer profile from Any to Left etc)

Also, if a profile is default, then the fields are "editable" assuming above bug is fixed. And it is not clear this is a default profile that you should clone first, etc. 

![image](https://user-images.githubusercontent.com/25975362/66330145-acc09300-e8e4-11e9-9939-85b82f2578fd.png)

![image](https://user-images.githubusercontent.com/25975362/66330394-2789ae00-e8e5-11e9-87cb-5203e6200f27.png)

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +143,267 @@ public override void OnInspectorGUI()
+        {
+            EditorGUILayout.Space();
+
+            SearchConfig config = new SearchConfig()
```
This whole session state and population tracking might honestly be worth it's own helper class

Troy-Ferrell replied:
SearchConfigPreferences or SearchConfigInspector or some other better name/utility because I am terrible at naming

Railboy replied:
Good idea.

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +143,267 @@ public override void OnInspectorGUI()
+                    bool selected = (config.SelectedSubjects & tag) != 0;
+                    GUI.color = selected ? MixedRealityInspectorUtility.SuccessColor : MixedRealityInspectorUtility.DisabledColor;
+                    if (GUILayout.Button(tag.ToString(), EditorStyles.miniButton))
+                    {
```
Obviously remove based on comments in main PR section. Nice added benefit is this will simplify the code up a lot too

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +143,267 @@ public override void OnInspectorGUI()
+            EditorGUILayout.Space();
+            EditorGUILayout.Space();
+
+            EditorGUILayout.HelpBox("Enter keywords below to search for fields.\nIf any subjects are selected, those will be used to narrow your search. Select (None) to search all fields.", MessageType.Info);
```
Switch to just bolded title/label "Search Fields" Or "Search Keywords" and if you want to be fancy add a search magnifying glass icon inline with the text box. I think there is InspectorUIUtility or MRStyleUtility?

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +143,267 @@ public override void OnInspectorGUI()
+            bool optionsFoldout = SessionState.GetBool(searchDisplayOptionsFoldoutKey, false);
+            optionsFoldout = EditorGUILayout.Foldout(optionsFoldout, "Options", true);
+            if (optionsFoldout)
+            {
```
This again could be part of that SearchConfigInspector class utility

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +143,267 @@ public override void OnInspectorGUI()
+
+            SessionState.SetString(searchDisplaySearchFieldKey, config.SearchFieldString);
+            SessionState.SetString(searchDisplaySelectedTagsKey, MixedRealitySearchUtility.SubjectFlagsToString(config.SelectedSubjects));
+            SessionState.SetBool(searchDisplayRequireAllTagsKey, config.RequireAllSubjects);
```
And this with SearchConfigInspector utility above...or again move better name :P

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +143,267 @@ public override void OnInspectorGUI()
+                                    EditorGUILayout.ObjectField(search.Profile, typeof(UnityEngine.Object), false, GUILayout.ExpandWidth(true));
+                                }
+
+                                EditorGUI.indentLevel++;
```
use => using new EditorGUI.IndentScope

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.FieldSearchResult.cs:
```diff
@@ -0,0 +1,16 @@
+namespace Microsoft.MixedReality.Toolkit.Utilities.Editor.Search
+{
+    /// <summary>
+    /// Struct for storing search results
```
Better comment. I know it's internal but each property also doesn't have comments.

Also maybe rename Property to Field? For some one new, SerializedProperty Property reads funny

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.SearchConfig.cs:
```diff
@@ -0,0 +1,22 @@
+namespace Microsoft.MixedReality.Toolkit.Utilities.Editor.Search
+{
+    /// <summary>
+    /// Struct for configuring a search.
```
Again better comments and comments for related items. How is this data container supposed to be used and what is it for?

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,307 @@
+                config.Keywords.RemoveWhere(s => s.Length < minSearchStringLength);
+            }
+
+            // The result that we will return, if not empty
```
remove comment, not useful.

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,307 @@
+                if (iterator.propertyType == SerializedPropertyType.ObjectReference && iterator.objectReferenceValue != null)
+                {   // If the property is itself a profile, perform a recursive search
+                    Type referenceType = iterator.objectReferenceValue.GetType();
+                    if (typeof(BaseMixedRealityProfile).IsAssignableFrom(referenceType))
```
Good place for Utility function => IsProfile or IsConcreteProfile and then remove comment

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,307 @@
+        /// <param name="config">Configuration for search.</param>
+        public static IEnumerable<ProfileSearchResult> SearchProfileFields(UnityEngine.Object profile, SearchConfig config)
+        {
+            if (string.IsNullOrEmpty(config.SearchFieldString) && config.SelectedSubjects == 0)
```
This is a really long function. Break this out into util functions etc. Some of this stuff could also be re-usable in MRInspectorUtility etc

Railboy replied:
I'll break it up once we know what we're doing with subject tags.

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,307 @@
+        /// <summary>
+        /// Converts a string flags value to a subject tag. Useful for storing in a session state string value.
+        /// </summary>
+        public static SubjectTag SubjectFlagsFromString(string subjectFlagValueString)
```
Probably something useful in the SearchConfigInspector

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.FieldSearchResult.cs:
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEditor;
```
@julenka/@davidkline-ms and other MRTK people, not sure how the team feels about MRSearchUtility.FieldSearchResult as a file name. 

Feels like it should just be under the MRSearchUtility folder with everything else

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,307 @@
+        /// <summary>
+        /// Field types that don't need their child properties displayed.
+        /// </summary>
+        private static readonly HashSet<string> serializedPropertyTypesToFlatten = new HashSet<string>()
```
If we switch to a simpler, render the profile if a match was found. This should be removed I believe?

That approach should also be faster since we can bail early if any match was found and just proceed to rendering

Railboy commented:
> It would be awesome if we could render the sub-profile and then bold or mark in red highlight the fields that match, but not sure how we could do that easily. 

@Troy-Ferrell Yes this would be _ideal_. Showing a list of relevant profiles was something we tried, but it turned out to be more frustrating than illuminating. As you said the problem was knowing _where_ in the profile to look.

So far my R&D hasn't turned up a way to do highlighting that doesn't compromise custom inspectors. We can iterate through serialized properties and draw a 'default' inspector with highlighting, but displaying two visually distinct versions of the profile is more confusing than helpful. Any ideas?

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+
+                    if (config.SearchTooltips)
+                    {
+                        matches |= property.tooltip.ToLower().Contains(keyword);
```
Same on to lower here

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+
+                // Check to see if this is a profile - if it is, enter the profile and check its children
+                bool isProfileField = false;
+                if (iterator.propertyType == SerializedPropertyType.ObjectReference && iterator.objectReferenceValue != null)
```
Probably easier to extract this out to IsProfileField() helper function and can combine if statements

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+        /// </summary>
+        /// <param name="profile">The configuration profile you're searching.</param>
+        /// <param name="config">Configuration for search.</param>
+        public static IEnumerable<ProfileSearchResult> SearchProfileFields(UnityEngine.Object profile, SearchConfig config)
```
This function is assuming it's going to analyze a MRTK profile. Make the profile argument of type BaseMixedRealityProfile to be explicit

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+                yield return result;
+            }
+
+            yield break;
```
necessary?

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+                            while (hasChildProperties)
+                            {
+                                // Check to see if this is a profile - if it is, enter the profile and check its children
+                                if (childIterator.propertyType == SerializedPropertyType.ObjectReference && childIterator.objectReferenceValue != null)
```
again, re-use the isProfileField() helper function here

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+                                    Type referenceType = childIterator.objectReferenceValue.GetType();
+                                    if (typeof(BaseMixedRealityProfile).IsAssignableFrom(referenceType))
+                                    {
+                                        isProfileField = true;
```
necessary?

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+                            SerializedProperty childIterator = iterator.Copy();
+                            bool hasChildProperties = childIterator.NextVisible(true);
+                            int depthOnEnter = childIterator.depth;
+                            while (hasChildProperties)
```
what are children of a SerializedProperty?

Also when searching the children, we are only looking for profiles?

Railboy replied:
Array values, members in a struct, etc.

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealitySearchUtility.cs:
```diff
@@ -0,0 +1,225 @@
+                CheckFieldKeywords(config, iterator, ref fieldMatchStrength, ref keywordReqsMet);
+                if (keywordReqsMet)
+                {
+                    if (ProfileSearchResult.IsEmpty(result))
```
the isempty here is a bit over-complicated. All you really need to track is the MaxFieldMatchStrength and a List<FieldSearchResult>. Why not declare these and then if your list is not empty at the end of the function, you create a ProfileSearchResult, populate, and return? 

This way you don't need to constantly check if isEmpty and can actually remove the function. You don't actually ever return an empty result so it's not utilized by the consumer of this utility either.

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +138,177 @@ public override void OnInspectorGUI()
+
+                SessionState.SetString(searchDisplaySearchFieldKey, config.SearchFieldString);
+                SessionState.SetBool(searchDisplayRequireAllKeywordsKey, config.RequireAllKeywords);
+                SessionState.SetBool(searchDisplaySearchTooltipsKey, config.SearchTooltips);
```
These should also be tied to the SearchConfigInspector

Troy-Ferrell commented on Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs:
```diff
@@ -118,11 +138,177 @@ public override void OnInspectorGUI()
+            if (optionsFoldout)
+            {
+                config.RequireAllKeywords = EditorGUILayout.Toggle("Require All Keywords", config.RequireAllKeywords);
+                config.SearchTooltips = EditorGUILayout.Toggle("Search Tooltips", config.SearchTooltips);
```
No tooltips. Not exaclty clear what child properties and gameobject names are considered

Railboy commented:
>The search performance is a bit weak. In particular, things slow down greatly if you type more than two words and you can see the keyboard input lagging.

I'll make it an async Task.

Railboy commented:
I've moved the search interface into a separate utility and loosely tied it to base profile inspector class. This made it simple to add search capabilities to profiles in general. I've updated our profile inspector classes to stop rendering if we're displaying search results. Here's a pointer profile selected in the project view:

![PointerProfileSearch](https://user-images.githubusercontent.com/9789716/66680938-82ddd800-ec26-11e9-82f1-b4141299b9f8.PNG)

![PointerProfile](https://user-images.githubusercontent.com/9789716/66680945-85d8c880-ec26-11e9-8b3c-2b62f872b106.PNG)

Searches have been made asynchronous for better performance.

@Troy-Ferrell @julenka Thanks for pushing this to be better - your feedback has improved this feature a great deal since the first iteration.

Railboy commented:
/azp run

Railboy commented:
Fixed merge conflict with dev branch, now works with the new dropdowns.

keveleigh commented:
/azp run

Railboy commented:
/azp run

Railboy commented:
Fixed merge conflict with dev branch.

Would be nice to get this merged soon, it's beginning to grow stale.

Railboy commented:
/azp run
